flink-runtime/src/main/java/org/apache/flink/runtime/state/AbstractTaskLocalStateStore.java

```diff
@@ -0,0 +1,430 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.state;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.api.common.JobID;
+import org.apache.flink.core.fs.FileSystem;
+import org.apache.flink.core.fs.Path;
+import org.apache.flink.runtime.checkpoint.TaskStateSnapshot;
+import org.apache.flink.runtime.clusterframework.types.AllocationID;
+import org.apache.flink.runtime.jobgraph.JobVertexID;
+import org.apache.flink.util.ExceptionUtils;
+import org.apache.flink.util.FlinkRuntimeException;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.AbstractMap;
+import java.util.AbstractMap.SimpleEntry;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.function.LongPredicate;
+
+/** Abstract TaskLocalStateStore. */
+public abstract class AbstractTaskLocalStateStore implements OwnedTaskLocalStateStore {
```

Comment:
Could you explain why a separate `AbstractTaskLocalStateStore` class, not just `ChangelogTaskLocalStateStore` extending `TaskLocalStateStoreImpl`?

And have you considered encapsulation instead of inheritance? Would it be more complex?

Reply:
Because I thought they were two different implementations before. Yes, `AbstractTaskLocalStateStore` is not necessary, `ChangelogTaskLocalStateStore` can extend `TaskLocalStateStoreImpl` directly, I would revert it later.
 
> And have you considered encapsulation instead of inheritance? Would it be more complex?

Yes,  because `ChangelogTaskLocalStateStore` differs `TaskLocalStateStoreImpl` in `storeLocalState()`, `confirmCheckpoint()` and `abortCheckpoint()`, which needs to use some **private method** of `TaskLocalStateStoreImpl`, so I choosed inheritance instead of  encapsulation.